File: anisotropic_hilbert.py
"""
anisotropic_hilbert.py

A continuous (lattice-continuous) Hilbert-like space-filling curve on an
axis-aligned dyadic box 2^{m0} × ... × 2^{m_{n-1}}.

This implementation follows the geometric machinery in:
  Chris Hamilton, "Compact Hilbert Indices", Technical Report CS-2006-07 (2006)

Key primitives used exactly as in the report:
  - T-transform:   T(e,d)(b) = (b XOR e) rotated-right by (d+1)        (Lemma 2.?? / Sec 2.1.3)
  - HilbertIndex update: e <- e XOR ( e(w) rotated-left by (d+1) ); d <- d + d(w) + 1 (Algorithm 2)
  - Entry points e(i) and directions d(i) sequences (Theorem 2.10 / Theorem 2.9)

We adapt these to "unequal dimensions" by:
  - Processing levels s = max(m) ... 1
  - At level s, the active axes are those with m_j >= s (monotone activation).
  - The number of active axes k(s) increases only when new dimensions "activate".
  - Hilbert digits are variable-width: k(s) bits at level s.
  - When k increases, we *embed* the current (e,d) into the larger active set by:
        e_new: copy old bits into their new positions; newly activated axes get 0-bit entry
        d_new: reindex d by axis-label (keep the *same physical axis* as the exit direction)

The resulting index space has exactly sum(m_j) bits, and the induced order of
points is a Hamiltonian path on the integer lattice grid: successive indices
map to points with Manhattan distance 1.

Public API:
  - encode(point, m):    point -> integer Hilbert index
  - decode(h, m):        integer Hilbert index -> point
  - iter_points(m):      iterate points in curve order

All coordinates and exponents are non-negative integers.
"""

from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from typing import Dict, Iterable, List, Sequence, Tuple


def gray_code(i: int) -> int:
    """Binary reflected Gray code: gc(i) = i XOR (i >> 1)."""
    return i ^ (i >> 1)


def gray_decode(g: int) -> int:
    """Inverse of binary reflected Gray code (gc^{-1})."""
    i = 0
    while g:
        i ^= g
        g >>= 1
    return i


def trailing_ones(i: int) -> int:
    """Number of trailing 1-bits in i (Hamilton's tsb)."""
    c = 0
    while i & 1:
        c += 1
        i >>= 1
    return c


def _mask(bits: int) -> int:
    return (1 << bits) - 1 if bits > 0 else 0


def rrotate(x: int, r: int, bits: int) -> int:
    """Rotate-right an unsigned 'bits'-bit word."""
    if bits <= 0:
        return x
    r %= bits
    x &= _mask(bits)
    if r == 0:
        return x
    return ((x >> r) | ((x & ((1 << r) - 1)) << (bits - r))) & _mask(bits)


def lrotate(x: int, r: int, bits: int) -> int:
    """Rotate-left an unsigned 'bits'-bit word."""
    if bits <= 0:
        return x
    r %= bits
    x &= _mask(bits)
    if r == 0:
        return x
    return ((x << r) | (x >> (bits - r))) & _mask(bits)


def T(e: int, d: int, bits: int, b: int) -> int:
    """
    Hamilton T-transform:
      T(e,d)(b) = (b XOR e) rotated-right by (d+1) (mod bits)
    """
    if bits <= 0:
        raise ValueError("bits must be positive")
    return rrotate((b ^ e) & _mask(bits), (d + 1) % bits, bits)


def T_inv(e: int, d: int, bits: int, b: int) -> int:
    """
    Inverse of T(e,d). Since T is (XOR e) then rotate-right, the inverse is:
      rotate-left by (d+1), then XOR e.
    """
    if bits <= 0:
        raise ValueError("bits must be positive")
    return (lrotate(b & _mask(bits), (d + 1) % bits, bits) ^ e) & _mask(bits)


@lru_cache(None)
def e_seq(bits: int, i: int) -> int:
    """
    Hamilton entry-point sequence e(i) for a k=bits dimensional cube.
    Theorem 2.10: e(0)=0; e(i)=gc(2*floor((i-1)/2)) for i>0.
    """
    if i == 0:
        return 0
    return gray_code(2 * ((i - 1) // 2))


@lru_cache(None)
def d_seq(bits: int, i: int) -> int:
    """
    Hamilton intra-subcube direction sequence d(i) for k=bits dimensions.
    Theorem 2.9 / Lemma 2.8:
      d(0)=0
      d(i)=g(i-1) if i even
      d(i)=g(i)   if i odd
    where g(i) = trailing_ones(i) and all indices are reduced mod bits.
    """
    if bits <= 0:
        raise ValueError("bits must be positive")
    if i == 0:
        return 0
    if i & 1:
        return trailing_ones(i) % bits
    return trailing_ones(i - 1) % bits


def _priority_order(m: Sequence[int]) -> List[int]:
    """Axis order used for activation embedding: smaller exponent first; tie by axis id."""
    return sorted(range(len(m)), key=lambda ax: (m[ax], ax))


def active_axes_by_level(m: Sequence[int]) -> Dict[int, List[int]]:
    """
    For levels s=1..mmax, return the ordered list of active axes:
      A_s = { axis j | m_j >= s } ordered by the priority rule.

    Monotonicity: A_s ⊆ A_{s-1}. (More axes activate as s decreases.)
    """
    m = tuple(m)
    if any(mi < 0 for mi in m):
        raise ValueError("All exponents m_j must be >= 0")
    n = len(m)
    mmax = max(m) if n else 0
    order = _priority_order(m)
    return {s: [ax for ax in order if m[ax] >= s] for s in range(1, mmax + 1)}


def embed_state(A_old: List[int], A_new: List[int], e_old: int, d_old: int) -> Tuple[int, int]:
    """
    Embed (e_old,d_old) defined on A_old into a larger active axis list A_new.

    - Bits of e_old are copied into the new positions of the same axes.
    - Newly activated axes get entry-bit 0.
    - d_old is reindexed by axis-label: keep the same *physical* direction axis.
    """
    k_old = len(A_old)
    k_new = len(A_new)
    if k_new < k_old:
        raise ValueError("A_new must be a superset of A_old")

    pos_new = {ax: j for j, ax in enumerate(A_new)}

    e_new = 0
    for j, ax in enumerate(A_old):
        if (e_old >> j) & 1:
            e_new |= 1 << pos_new[ax]

    dir_axis = A_old[d_old]  # axis-label in global coordinates
    d_new = pos_new[dir_axis]
    return e_new, d_new


def index_bit_length(m: Sequence[int]) -> int:
    """Total number of bits in the (variable-width) index. For dyadic boxes this equals sum(m)."""
    return sum(int(mi) for mi in m)


def encode(point: Sequence[int], m: Sequence[int]) -> int:
    """
    Map an n-D integer lattice point to its anisotropic Hilbert index.

    Parameters
    ----------
    point : sequence of ints
        Coordinates (p0,...,p_{n-1}) with 0 <= pj < 2^{m_j}.
    m : sequence of ints
        Exponents (m0,...,m_{n-1}) describing the box size 2^{m0}×...×2^{m_{n-1}}.

    Returns
    -------
    h : int
        Hilbert index in [0, 2^{sum(m)}).

    Notes
    -----
    This is the Hamilton HilbertIndex algorithm adapted to variable-width digits:
    at level s, the digit width is k(s) = |{j: m_j >= s}|.
    """
    m = tuple(int(mi) for mi in m)
    point = tuple(int(x) for x in point)
    if len(point) != len(m):
        raise ValueError("point and m must have the same length")
    if any(mi < 0 for mi in m):
        raise ValueError("All exponents m_j must be >= 0")

    n = len(m)
    mmax = max(m) if n else 0
    if mmax == 0:
        return 0

    # bounds check
    for j in range(n):
        if not (0 <= point[j] < (1 << m[j] if m[j] else 1)):
            raise ValueError(f"point[{j}] out of bounds for m[{j}]={m[j]}")

    axes_by_s = active_axes_by_level(m)

    h = 0
    e = 0
    d = 0

    for s in range(mmax, 0, -1):
        A = axes_by_s[s]
        k = len(A)

        e &= _mask(k)
        d %= k

        # Pack the s-1 bit across active axes into an integer l (k bits).
        l = 0
        for j, ax in enumerate(A):
            l |= ((point[ax] >> (s - 1)) & 1) << j

        # Hamilton step: transform to standard orientation, Gray-decode to get w.
        l_t = T(e, d, k, l)
        w = gray_decode(l_t)

        # Append this variable-width digit.
        h = (h << k) | w

        # Update orientation using the cube's e(w), d(w).
        e = e ^ lrotate(e_seq(k, w), (d + 1) % k, k)
        d = (d + d_seq(k, w) + 1) % k

        # Activation embedding to next level if the active set grows.
        if s > 1:
            A_next = axes_by_s[s - 1]
            if len(A_next) > k:
                e, d = embed_state(A, A_next, e, d)

    return h


def decode(h: int, m: Sequence[int]) -> Tuple[int, ...]:
    """
    Inverse of encode(): map an index to its n-D lattice point.

    Parameters
    ----------
    h : int
        Hilbert index in [0, 2^{sum(m)}).
    m : sequence of ints
        Exponents (m0,...,m_{n-1}).

    Returns
    -------
    point : tuple[int,...]
        Coordinates (p0,...,p_{n-1}).
    """
    m = tuple(int(mi) for mi in m)
    if any(mi < 0 for mi in m):
        raise ValueError("All exponents m_j must be >= 0")

    n = len(m)
    mmax = max(m) if n else 0
    if mmax == 0:
        return tuple([0] * n)

    M = index_bit_length(m)
    if not (0 <= h < (1 << M)):
        raise ValueError(f"h must be in [0, 2^{M})")

    axes_by_s = active_axes_by_level(m)

    # Segment sizes are k(s) for s=mmax..1
    seg_sizes = [len(axes_by_s[s]) for s in range(mmax, 0, -1)]
    bit_pos = sum(seg_sizes)

    e = 0
    d = 0
    point = [0] * n

    for s in range(mmax, 0, -1):
        A = axes_by_s[s]
        k = len(A)

        bit_pos -= k
        w = (h >> bit_pos) & _mask(k)

        # Hamilton inverse step: Gray-encode w, untransform back to current orientation.
        l = gray_code(w)
        l = T_inv(e, d, k, l)

        # Write the recovered bit into each active axis coordinate.
        for j, ax in enumerate(A):
            point[ax] |= ((l >> j) & 1) << (s - 1)

        # Update orientation and embed if activation grows.
        e = e ^ lrotate(e_seq(k, w), (d + 1) % k, k)
        d = (d + d_seq(k, w) + 1) % k

        if s > 1:
            A_next = axes_by_s[s - 1]
            if len(A_next) > k:
                e, d = embed_state(A, A_next, e, d)

    return tuple(point)


def iter_points(m: Sequence[int]) -> Iterable[Tuple[int, ...]]:
    """Iterate all points in curve order (decode 0..2^{sum(m)}-1)."""
    m = tuple(int(mi) for mi in m)
    M = index_bit_length(m)
    for h in range(1 << M):
        yield decode(h, m)

File: anisotropic_hilbert.c
/*
 * anisotropic_hilbert.c
 *
 * A continuous (lattice-continuous) Hilbert-like space-filling curve on an
 * axis-aligned dyadic box 2^{m0} × ... × 2^{m_{n-1}}.
 *
 * This implementation follows the geometric machinery in:
 *   Chris Hamilton, "Compact Hilbert Indices", Technical Report CS-2006-07 (2006)
 *
 * Supports up to MAX_DIMS dimensions and uses __uint128_t for Hilbert indices
 * to handle large sum(m) values.
 */

#include <stdint.h>
#include <string.h>
#include <stdbool.h>

#define MAX_DIMS 32

/* Use 128-bit integers for Hilbert indices (GCC/Clang extension) */
typedef __uint128_t hindex_t;
typedef uint32_t coord_t;

/* --------------------------------------------------------------------------
 * Bit manipulation primitives
 * -------------------------------------------------------------------------- */

static inline uint32_t gray_code(uint32_t i) {
    return i ^ (i >> 1);
}

static inline uint32_t gray_decode(uint32_t g) {
    uint32_t i = 0;
    while (g) {
        i ^= g;
        g >>= 1;
    }
    return i;
}

static inline int trailing_ones(uint32_t i) {
    int c = 0;
    while (i & 1) {
        c++;
        i >>= 1;
    }
    return c;
}

static inline uint32_t mask_bits(int bits) {
    return bits > 0 ? ((1U << bits) - 1) : 0;
}

static inline uint32_t rrotate(uint32_t x, int r, int bits) {
    if (bits <= 0) return x;
    r = r % bits;
    x &= mask_bits(bits);
    if (r == 0) return x;
    return ((x >> r) | ((x & ((1U << r) - 1)) << (bits - r))) & mask_bits(bits);
}

static inline uint32_t lrotate(uint32_t x, int r, int bits) {
    if (bits <= 0) return x;
    r = r % bits;
    x &= mask_bits(bits);
    if (r == 0) return x;
    return ((x << r) | (x >> (bits - r))) & mask_bits(bits);
}

/* Hamilton T-transform: T(e,d)(b) = (b XOR e) rotated-right by (d+1) */
static inline uint32_t T_transform(uint32_t e, int d, int bits, uint32_t b) {
    return rrotate((b ^ e) & mask_bits(bits), (d + 1) % bits, bits);
}

/* Inverse of T(e,d) */
static inline uint32_t T_inv(uint32_t e, int d, int bits, uint32_t b) {
    return (lrotate(b & mask_bits(bits), (d + 1) % bits, bits) ^ e) & mask_bits(bits);
}

/* Entry-point sequence e(i) */
static inline uint32_t e_seq(int i) {
    if (i == 0) return 0;
    return gray_code(2 * ((i - 1) / 2));
}

/* Direction sequence d(i) */
static inline int d_seq(int bits, int i) {
    if (i == 0) return 0;
    if (i & 1) return trailing_ones(i) % bits;
    return trailing_ones(i - 1) % bits;
}

/* --------------------------------------------------------------------------
 * Axis ordering and active axes
 * -------------------------------------------------------------------------- */

/* Comparison function for sorting axes by (m[ax], ax) */
typedef struct {
    int axis;
    int exp;
} axis_exp_t;

static void sort_axes_by_exp(const int *m, int n, int *order) {
    axis_exp_t items[MAX_DIMS];
    for (int i = 0; i < n; i++) {
        items[i].axis = i;
        items[i].exp = m[i];
    }
    /* Simple insertion sort (n is small) */
    for (int i = 1; i < n; i++) {
        axis_exp_t key = items[i];
        int j = i - 1;
        while (j >= 0 && (items[j].exp > key.exp ||
               (items[j].exp == key.exp && items[j].axis > key.axis))) {
            items[j + 1] = items[j];
            j--;
        }
        items[j + 1] = key;
    }
    for (int i = 0; i < n; i++) {
        order[i] = items[i].axis;
    }
}

/* Get active axes at level s, returns count */
static int get_active_axes(const int *m, int n, int s, const int *order, int *active) {
    int k = 0;
    for (int i = 0; i < n; i++) {
        int ax = order[i];
        if (m[ax] >= s) {
            active[k++] = ax;
        }
    }
    return k;
}

/* Embed state from smaller to larger active set */
static void embed_state(const int *A_old, int k_old, const int *A_new, int k_new,
                        uint32_t e_old, int d_old, uint32_t *e_new, int *d_new) {
    /* Build position map for new axes */
    int pos_new[MAX_DIMS];
    for (int j = 0; j < k_new; j++) {
        pos_new[A_new[j]] = j;
    }

    /* Copy e bits to new positions */
    *e_new = 0;
    for (int j = 0; j < k_old; j++) {
        if ((e_old >> j) & 1) {
            *e_new |= 1U << pos_new[A_old[j]];
        }
    }

    /* Reindex d by axis label */
    int dir_axis = A_old[d_old];
    *d_new = pos_new[dir_axis];
}

/* --------------------------------------------------------------------------
 * Main encode/decode functions
 * -------------------------------------------------------------------------- */

/*
 * Encode a point to its Hilbert index.
 *
 * Parameters:
 *   point: array of n coordinates
 *   m: array of n exponents (box is 2^m[0] x ... x 2^m[n-1])
 *   n: number of dimensions
 *
 * Returns: Hilbert index
 */
hindex_t hilbert_encode(const coord_t *restrict point, const int *restrict m, int n) {
    if (n <= 0 || n > MAX_DIMS) return 0;

    int mmax = 0;
    for (int i = 0; i < n; i++) {
        if (m[i] > mmax) mmax = m[i];
    }
    if (mmax == 0) return 0;

    /* Precompute axis order */
    int order[MAX_DIMS];
    sort_axes_by_exp(m, n, order);

    /* Precompute active axes for each level */
    int active[MAX_DIMS][MAX_DIMS];
    int k_at_level[MAX_DIMS + 1];
    for (int s = 1; s <= mmax; s++) {
        k_at_level[s] = get_active_axes(m, n, s, order, active[s]);
    }

    hindex_t h = 0;
    uint32_t e = 0;
    int d = 0;

    for (int s = mmax; s >= 1; s--) {
        int *A = active[s];
        int k = k_at_level[s];

        e &= mask_bits(k);
        d = d % k;

        /* Pack the (s-1) bit across active axes */
        uint32_t l = 0;
        for (int j = 0; j < k; j++) {
            int ax = A[j];
            l |= ((point[ax] >> (s - 1)) & 1) << j;
        }

        /* Hamilton step */
        uint32_t l_t = T_transform(e, d, k, l);
        uint32_t w = gray_decode(l_t);

        /* Append digit */
        h = (h << k) | w;

        /* Update orientation */
        e = e ^ lrotate(e_seq(w), (d + 1) % k, k);
        d = (d + d_seq(k, w) + 1) % k;

        /* Embed if active set grows */
        if (s > 1 && k_at_level[s - 1] > k) {
            uint32_t e_new;
            int d_new;
            embed_state(A, k, active[s - 1], k_at_level[s - 1], e, d, &e_new, &d_new);
            e = e_new;
            d = d_new;
        }
    }

    return h;
}

/*
 * Decode a Hilbert index to its point.
 *
 * Parameters:
 *   h: Hilbert index
 *   m: array of n exponents
 *   n: number of dimensions
 *   point: output array of n coordinates
 */
void hilbert_decode(hindex_t h, const int *restrict m, int n, coord_t *restrict point) {
    memset(point, 0, n * sizeof(coord_t));

    if (n <= 0 || n > MAX_DIMS) return;

    int mmax = 0;
    int total_bits = 0;
    for (int i = 0; i < n; i++) {
        if (m[i] > mmax) mmax = m[i];
        total_bits += m[i];
    }
    if (mmax == 0) return;

    /* Precompute axis order */
    int order[MAX_DIMS];
    sort_axes_by_exp(m, n, order);

    /* Precompute active axes for each level */
    int active[MAX_DIMS][MAX_DIMS];
    int k_at_level[MAX_DIMS + 1];
    for (int s = 1; s <= mmax; s++) {
        k_at_level[s] = get_active_axes(m, n, s, order, active[s]);
    }

    /* Compute segment sizes and bit position */
    int bit_pos = total_bits;

    uint32_t e = 0;
    int d = 0;

    for (int s = mmax; s >= 1; s--) {
        int *A = active[s];
        int k = k_at_level[s];

        bit_pos -= k;
        uint32_t w = (uint32_t)((h >> bit_pos) & mask_bits(k));

        /* Hamilton inverse step */
        uint32_t l = gray_code(w);
        l = T_inv(e, d, k, l);

        /* Write recovered bits */
        for (int j = 0; j < k; j++) {
            int ax = A[j];
            point[ax] |= ((l >> j) & 1) << (s - 1);
        }

        /* Update orientation */
        e = e ^ lrotate(e_seq(w), (d + 1) % k, k);
        d = (d + d_seq(k, w) + 1) % k;

        /* Embed if active set grows */
        if (s > 1 && k_at_level[s - 1] > k) {
            uint32_t e_new;
            int d_new;
            embed_state(A, k, active[s - 1], k_at_level[s - 1], e, d, &e_new, &d_new);
            e = e_new;
            d = d_new;
        }
    }
}

/* --------------------------------------------------------------------------
 * Batch operations for efficiency
 * -------------------------------------------------------------------------- */

/*
 * Decode multiple consecutive Hilbert indices starting from h_start.
 * Writes n_points * n coordinates to points array (row-major).
 */
void hilbert_decode_batch(hindex_t h_start, uint64_t n_points,
                          const int *restrict m, int n,
                          coord_t *restrict points) {
    for (uint64_t i = 0; i < n_points; i++) {
        hilbert_decode(h_start + i, m, n, points + i * n);
    }
}

/*
 * Encode multiple points.
 * points is row-major array of n_points * n coordinates.
 * indices is output array of n_points Hilbert indices.
 */
void hilbert_encode_batch(const coord_t *restrict points, uint64_t n_points,
                          const int *restrict m, int n,
                          hindex_t *restrict indices) {
    for (uint64_t i = 0; i < n_points; i++) {
        indices[i] = hilbert_encode(points + i * n, m, n);
    }
}

/* --------------------------------------------------------------------------
 * C-compatible wrapper functions for ctypes (64-bit index versions)
 *
 * Since ctypes doesn't directly support __uint128_t, we provide 64-bit
 * versions that work for sum(m) <= 64. For larger indices, use the
 * hi/lo split versions.
 * -------------------------------------------------------------------------- */

uint64_t hilbert_encode_64(const coord_t *point, const int *m, int n) {
    return (uint64_t)hilbert_encode(point, m, n);
}

void hilbert_decode_64(uint64_t h, const int *m, int n, coord_t *point) {
    hilbert_decode((hindex_t)h, m, n, point);
}

/* 128-bit versions using hi/lo split */
void hilbert_encode_128(const coord_t *point, const int *m, int n,
                        uint64_t *h_lo, uint64_t *h_hi) {
    hindex_t h = hilbert_encode(point, m, n);
    *h_lo = (uint64_t)h;
    *h_hi = (uint64_t)(h >> 64);
}

void hilbert_decode_128(uint64_t h_lo, uint64_t h_hi, const int *m, int n,
                        coord_t *point) {
    hindex_t h = ((hindex_t)h_hi << 64) | h_lo;
    hilbert_decode(h, m, n, point);
}

/* Batch versions (64-bit) */
void hilbert_decode_batch_64(uint64_t h_start, uint64_t n_points,
                             const int *m, int n, coord_t *points) {
    hilbert_decode_batch((hindex_t)h_start, n_points, m, n, points);
}

void hilbert_encode_batch_64(const coord_t *points, uint64_t n_points,
                             const int *m, int n, uint64_t *indices) {
    for (uint64_t i = 0; i < n_points; i++) {
        indices[i] = (uint64_t)hilbert_encode(points + i * n, m, n);
    }
}

/* Utility: compute total index bits */
int hilbert_index_bits(const int *m, int n) {
    int total = 0;
    for (int i = 0; i < n; i++) {
        total += m[i];
    }
    return total;
}
